### PR TITLE
WIP: implement migration

### LINKED ISSUE #2: add `jupyter migrate`
explicit command for migrating from an IPython profile to Jupyter

### PR DESCRIPTION
**Pull request by minrk**
Currently, this implements migration of config files and user-installed nbextensions, kernelspecs from IPython < 4.0 to Jupyter locations via `jupyter migrate` as:

- copy files to new locations
- if new locations already exist, do nothing
- never overwrite anything
- don't copy empty config files
- small amount of find-replace in config files for renamed classes (e.g. IPythonQtConsoleApp -> JupyterQtConsoleApp)

To figure out for this PR:

- [x] what files should be migrated
- [x] copy or move, etc.? Allow overwrite?
- [x] API for migration (CLI)
- [x] should there be any modifications to contents, or only copies?

To figure out after this PR:

- When to migrate.

I want to implement migration first, and work on implicit-on-startup logic, if any, as a separate discussion.

cc @takluyver @fperez @ellisonbg

closes #2

### CODE REVIEW

**Comment by Carreau** [comment]
> copy files to new locations

I would move it, but copy is fine. 
The reason to move is that if file still exist in old location, you will be tempted to edit it. 
While an error make you think (most of the time). 

devil advocate, some people might stil have old versions of IPython installed ?

> if new locations already exist, do nothing

Seem legit. 

> never overwrite anything

+1

> don't copy empty config files

+1 , I suppose you consider no whitespace and comment as irrelevant. 

> small amount of find-replace in config files for renamed classes (e.g. IPythonQtConsoleApp -> JupyterQtConsoleApp)

Seem legit. 

> what files should be migrated

Profile default. Warn if other profiles ?

> copy or move, etc.? Allow overwrite?

Move (said above), no overwrite, one can nuke it's jupyter-config folder and relaunch migration if necessary. 

> API for migration (CLI)

except `jupyter migrate`, I 'm not sure what we should add. For a root user to be able to migrate each users ?

> should there be any modifications to contents, or only copies?

Minimal modif IMHO.

> When to migrate.

At Jupyter/IPython startup, if no config file, look for old IPython place, if something die and ask to run jupyter migrate. No magic involve.

**Comment by minrk** [comment]
> devil advocate, some people might still have old versions of IPython installed ?

I know I will have 3.x running for some time for occasional testing. It would probably be annoying if every time I launched Jupyter, I lost all my IPython 2, 3 config.

> Move (said above), no overwrite, one can nuke it's jupyter-config folder and relaunch migration if necessary.

This actually argues for copy as well, unless I'm misunderstanding. If we move files, the originals don't exist, so nuking the new config dir means destroying the only copy, in which case relaunch will have nothing to migrate since the originals are gone.

> Profile default. Warn if other profiles ?

Agreed on profile default. Warning seems a bit much, but probably something in the output indicating that other profiles were not migrated.

> except jupyter migrate, I 'm not sure what we should add.

I'm mainly wondering if we want to have something granular, like allowing users to migrate one piece explicitly, e.g. `jupyter migrate --config` or `jupyter migrate --kernels`, etc. I'm expecting that's overkill, but I wanted to get feedback.

> At Jupyter/IPython startup

So run it at the outer-most `jupyter` command, before dispatching to subcommands? That makes sense to me. Do we want to write a `.migrated` marker file, to indicate that migration has already run, to avoid walking through a bunch of files at every invocation?

**Comment by Carreau** [comment]
> So run it at the outer-most jupyter command, before dispatching to subcommands? That makes sense to me. Do we want to write a .migrated marker file, to indicate that migration has already run, to avoid walking through a bunch of files at every invocation?

Technically you try to migrate only if you don't find your own config, no ? So there is no overhead.

**Comment by minrk** [comment]
Okay, that's what I misunderstood. And that would mean migration definitely is *not* handled by the outer application, but by every individual application.

It would mean that migration runs every time if the user has no config, though, which is the majority case, I think.

**Comment by Carreau** [comment]
> It would mean that migration runs every time if the user has no config, though, which is the majority case, I think.

Ah, true in Jupyter config is empty by default I forgot that.

File: jupyter_core/migrate.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+import shutil
+
+from traitlets.config import PyFileConfigLoader, JSONFileConfigLoader
+
+from ipython_genutils.path import ensure_dir_exists
+try:
+    from IPython.paths import get_ipython_dir, locate_profile
+except ImportError:
+    # IPython < 4
+    from IPython.utils.path import get_ipython_dir, locate_profile
```
Comment (Carreau) [inline jupyter_core/migrate.py]: Will we ever by IPython <4 ? If so should we just abort  ?

Comment (Carreau) [inline jupyter_core/migrate.py]: Or at least warn?

File: jupyter_core/migrate.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    dst = pjoin('{jupyter_data}', security_file)
+    migrations[src] = dst
+
+config_migrations = ['notebook', 'nbconvert', 'qtconsole']
+
+regex = re.compile
+
+config_substitutions = {
+    regex(r'\bIPythonQtConsoleApp\b'): 'JupyterQtConsoleApp',
+    regex(r'\bIPython\.html\b'): 'jupyter_notebook',
```
Comment (Carreau) [inline jupyter_core/migrate.py]: `notebook`

File: jupyter_core/migrate.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    regex(r'\bIPython\.nbconvert\b'): 'jupyter_nbconvert',
```
Comment (Carreau) [inline jupyter_core/migrate.py]: `nbconvert`

**Comment by Carreau** [comment]
2 replacement since we re-renamed the repos. But +1

**Comment by minrk** [comment]
@Carreau thanks, fixed the package names. Should be good to go.

**Pull request merged**